### PR TITLE
 osconfig: support rebootConfig, windows update classifications and excludes

### PR DESCRIPTION
**Pull request by adjackura**
Write log to guest attributes, clean up locking logic

### CODE REVIEW

File: cli_tools/osconfig_agent/updates_windows.go [hunk 1]
```diff
@@ -34,16 +38,190 @@ func rebootRequired() (bool, error) {
...
+	}
+	kbArticleIDsCount, _ := kbArticleIDsCountRaw.Value().(int32)
+
+	fmt.Printf("DEBUG: filtering out KBs: %q\n", excludes)
+	for i := 0; i < int(kbArticleIDsCount); i++ {
+		kbRaw, err := kbArticleIDs.GetProperty("Item", i)
+		if err != nil {
+			return err
+		}
+		if _, ok := excludes[kbRaw.ToString()]; ok {
```
Comment (ryanwe) [inline cli_tools/osconfig_agent/updates_windows.go]: consider separating the COM logic from the filter logic. Ideally, the filter logic would read cleanly with the COM logic abstracted away.

File: cli_tools/osconfig_agent/updates_windows.go [hunk 2]
```diff
@@ -34,16 +38,190 @@ func rebootRequired() (bool, error) {
...
+	defer updateCollObj.Release()
+
+	updateColl, err := updateCollObj.IDispatch(ole.IID_IDispatch)
+	if err != nil {
+		return err
+	}
+	defer updateColl.Release()
+
+	class := make(map[string]struct{})
+	for _, c := range pp.WindowsUpdate.Classifications {
```
Comment (ryanwe) [inline cli_tools/osconfig_agent/updates_windows.go]: is this additional filtering?

File: go/packages/packages_linux.go [hunk 3]
```diff
@@ -498,15 +497,15 @@ func installedGEM() ([]PkgInfo, error) {
...
+		fmt.Println("No gems installed.")
 		return nil, nil
 	}
 
 	var pkgs []PkgInfo
 	for _, ln := range lines[2:] {
 		pkg := strings.Fields(ln)
 		if len(pkg) < 2 {
-			logger.Errorf("%q does not represent a gem", ln)
+			fmt.Printf("%q does not represent a gem\n", ln)
```
Comment (ryanwe) [inline go/packages/packages_linux.go]: Might be a good time to add the beginnings of a custom logger even if the logger just calls `fmt.Printf()`

Comment (adjackura) [reply go/packages/packages_linux.go]: Did this for the main binary, these libraries still need to have something passed in which was a bit more than I wanted to setup now

**Pull request merged**